### ISSUE TITLE
Simpletest: Exceptions with htaccess files when tests run with profile caching

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4685]
**Description of the bug**

[Recently](https://github.com/backdrop/backdrop-issues/issues/706) the active and staging config directories got protection via htaccess. These files are causing problems when running tests with caching on, as the file permission is 444 and so only the first copy succeeds.

**Steps To Reproduce**

1. Make sure that test caching is turned on (admin/config/development/testing/settings)
2. Run a single test, for instance the first one (Actions executing in a potentially infinite loop)

**Actual behavior**

You get two Exceptions:

```
EXCEPTION: copy(files/simpletest/626731/config_active/.htaccess): failed to open stream: Permission denied	Warning	backdrop_web_test_case.php	1643	BackdropWebTestCase->recursiveCopy()
EXCEPTION: copy(files/simpletest/626731/config_staging/.htaccess): failed to open stream: Permission denied	Warning	backdrop_web_test_case.php	1643	BackdropWebTestCase->recursiveCopy()
```
Note: Copying happens in test setup, no test actually fails.

No clue, why this doesn't seem to be a problem with our testboxes. Can anyone shed light on it?

**Additional information**

- Backdrop CMS version: latest dev
- Web server: apache with php-fpm
- Reproducible with different PHP versions
- Reproducible when tests run via command line or via UI

**Comment by indigoxela** [comment]
Here's a [PR](https://github.com/backdrop/backdrop/pull/3358) for testing and review.

**Comment by indigoxela** [comment]
I just verified that the exceptions also happen in an (otherwise unrelated) testbox - via UI. So it's not just my server.

**Comment by bradbulger** [comment]
It creates the .htaccess with inaccessible permissions. Is just not copying it compatible with the changes from #706?

**Comment by bradbulger** [comment]
I confirmed that it happens with a generic demo sandbox installation.

**Comment by indigoxela** [comment]
> It creates the .htaccess with inaccessible permissions

The htaccess files intentionally have permission 444 (read only). But that causes problems with the copy() of them. The easiest solution is to skip copying these two files.

**Comment by klonos** [comment]
Code looks good 👍 ...I need to find some time to test this properly on my local (unless someone else beats me to it).

**Comment by bradbulger** [comment]
Works on my local copy.

**Comment by BWPanda** [comment]
The PR prevents copying any `.htaccess` file, but I wonder if we should specify only ignoring `.htaccess` files in the `/config` directories...?

**Comment by indigoxela** [comment]
> The PR prevents copying any .htaccess file, but I wonder if we should specify only ignoring .htaccess files in the /config directories

Can you describe a bit, what "ignoring" means in this context? Not copying it in function recursiveCopy() is sort of ignoring, but you seem to be after something else ... which I fail to understand. :wink:

**Comment by BWPanda** [comment]
Sorry, let me rephrase: The PR prevents copying *any* `.htaccess` file, but I wonder if we should just prevent it copying *only* `.htaccess` files in the `/config` directories...?

**Comment by indigoxela** [comment]
> I wonder if we should just prevent it copying only .htaccess files in the /config directories...?

Ah, now I get it, possibly a third function param in recursiveCopy - like "$ignore_htaccess", which only prevents copy() on it if set to true?

Hm. Currently that function is only called in that single test.

And Backdrop doesn't ship with / create so many htaccess files. There's the main one in the base directory, one per file directory (public/private) and one in the temporary dir.

When could copying these be necessary in any (future) test?

**Comment by BWPanda** [comment]
> Also note: chmod to 444 is run on all of the created htaccess files.

Ok. Not sure then. I just thought if we haven't been having a problem with `.htaccess` files before now, then maybe excluding all of them (rather than just the problematic ones) from that function might cause future issues...

**Comment by indigoxela** [comment]
> Ok. Not sure then. I just thought if we haven't been having a problem with .htaccess files before now

The two htaccess files in the config directories are rather new. Since they exist, this test setup chokes on them, if profile caching is on. Before that there wasn't any recursive copy of any folder containing such a file. In fact, this is the only usage of function recursiveCopy.

If you see a possible use case, or simply prefer it that way, I could of course add it.

**Comment by indigoxela** [comment]
@BWPanda did you decide anything here yet?

**Comment by BWPanda** [comment]
I was just thinking out loud. Ignore me :slightly_smiling_face:

**Comment by quicksketch** [comment]
Thanks @indigoxela for filing and fixing this issue! I've run into it a few times also and didn't get to filing an issue. Really appreciate @BWPanda, @bradbulger, and @klonos for reviewing and confirming this. Merged into 1.x and 1.17.x.
